### DISCUSSION TITLE
NPM-module vs. Deno.js library

### DISCUSSION THREAD

**Discussion by pubmikeb** [denoland/deno#16749] [General]
With introduction of the NPM-compatibility in Deno.js, a developer can chose wherever to use an NPM-module or a suitable Deno.js library, e.g.:
* [node-argon2](https://www.npmjs.com/package/argon2) as the NPM-package
* [Argon2 for Deno](https://deno.land/x/argon2@v0.9.2) as the Deno.js library

or
* [ESLint](https://www.npmjs.com/package/eslint)
* [Deno Linter](https://deno.land/manual@v1.28.1/tools/linter)

In case there are alternatives, which version should a developer prefer?
This question especially relevant for linters, where from one side we have ESLint — a proved and currently the most advanced JS linter with extremely wide community and active development cycle with releases about every two weeks, and from the other — Deno Linter with a limited ruleset.

P.S. By mistake, initially, I've opened an [issue](https://github.com/denoland/deno/issues/16746) on the subject.

**Comment by jtoppine** [comment]
(copying my earlier response from the initial issue to this discussion)

There's no question about it; for Deno, it is better to use a library specifically made for Deno. It will use modern, mostly standard APIs Deno is built around, so they likely have cleaner, more efficient codebase, and more "modern" API. Also the additional complexity of node compatibility layer is avoided.

However, like with any library, one needs to be mindfull of how well any given library is maintained and supported. Node has been around for a long time, and is very popular. Thus many npm libraries are more mature compared to the relatively young Deno libraries that may have some bugs, missing features, and smaller community of forward-looking developers doing the legwork of ironing all the small issues out.

So, I suggest choose the Deno library by default. If that is incomplete or doesn't work out for you for some reason, only then fall back to the NPM version.
